go-filter-chain/filters/jwt.go

```diff
@@ -0,0 +1,183 @@
+package filters
+
+import (
+	"context"
+	"crypto/rsa"
+	"errors"
+	"github.com/dgrijalva/jwt-go"
+	ext_authz "github.com/envoyproxy/go-control-plane/envoy/service/auth/v2"
+	envoy_type "github.com/envoyproxy/go-control-plane/envoy/type"
+	"github.com/gogo/googleapis/google/rpc"
+	"github.com/patrickmn/go-cache"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/genproto/googleapis/rpc/status"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+
+
+type TokenData struct {
+	meta_clientType        string
+	applicationConsumerKey string
+	applicationName        string
+	applicationId          string
+	applicationOwner       string
+	apiCreator             string
+	apiCreatorTenantDomain string
+	apiTier                string
+	username               string
+	userTenantDomain       string
+	throttledOut           bool
+	serviceTime            int64
+	authorized             bool
+}
+
+var (
+	Unknown = "__unknown__"
+	once sync.Once
+	once_1 sync.Once
+	caCert []byte
+	err error
+	key *rsa.PublicKey
+	jwtCache = cache.New(5*time.Minute, 10*time.Minute)
+	UnauthorizedError = errors.New("Invalid access token")
+	jwtToken string
+	isCacheEnabled bool
+)
+
+
+
+// handle JWT token
+func HandleJWT(validateSubscription bool, publicCert []byte, token string) (bool, TokenData, error) {
+
+	accessToken := token
+
+	tokenContent := strings.Split(accessToken, ".")
+	var tokenData TokenData
+
+	if len(tokenContent) != 3 {
+		log.Errorf("Invalid JWT token received, token must have 3 parts")
+		return false, tokenData, UnauthorizedError
+	}
+
+	signedContent := tokenContent[0] + "." + tokenContent[1]
+	err := validateSignature(publicCert, signedContent, tokenContent[2])
+	if err != nil {
+		log.Errorf("Error in validating the signature: %v", err)
+		return false, tokenData, UnauthorizedError
+	}
+
+	return true, tokenData, nil
+}
+
+// validate the signature
+func validateSignature(publicCert []byte, signedContent string, signature string) error {
+
+	once_1.Do(func() {
+		key, err = jwt.ParseRSAPublicKeyFromPEM(publicCert)
+		log.Info("read public key once")
+	})
+
+	if err != nil {
+		log.Errorf("Error in parsing the public key: %v", err)
+		return err
+	}
+
+	return jwt.SigningMethodRS256.Verify(signedContent, signature, key)
+}
+
+//reading the secret
+func ReadFile(fileName string) ([]byte, error) {
+
+	secretValue, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		log.Warnf("Error in reading the file %v: error - %v", fileName, err)
+	}
+
+	return secretValue, err
+}
+
+
+func ValidateToken(ctx context.Context, req *ext_authz.CheckRequest) (*ext_authz.CheckResponse, error) {
+
+	once.Do(func() {
+		caCert,_ = ReadFile("./artifacts/server.pem")
```

Comment:
read this only once. not for each request

Reply:
Once.Do function execute only once